### ISSUE TITLE
[BUG] Inconsistency between auth_db/get_ha1 and auth/auth_get_ha1 regarding username with domain

### ISSUE THREAD

**Issue by c-leroy** [OpenSIPS/opensips#2356]
**OpenSIPS version you are running**

```
version: opensips 2.4.9 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
main.c compiled on  with gcc 6.3.0
```

**Describe the bug**

We were using `www_authorize` from `auth_db` to authenticate `REGISTER`, and we are willing to use `pv_www_authorize` from `auth module`.
When a `REGISTER` message includes a username with domain in `username` field of `Authorization` header, it succeeds with `www_authorize`, which only takes the userpart.
With `pv_www_authorize`, it fails with invalid user.
It is because `auth_get_ha1()` takes whole username, while `get_ha1()` takes userpart.
It may come from commit 3b6685542b780ec47042a0e376de108e6d9c5fb7

**To Reproduce**

`REGISTER` succeeding with `www_authorize`, failing with `pv_www_authorize`
<pre>
REGISTER sip:sip.domain.net SIP/2.0.
Via: SIP/2.0/UDP 192.168.1.6:5060;branch=z9hG4bK.B5tpVQoaa;rport.
From: <sip:user@sip.domain.net>;tag=mwe37-5JQ.
To: sip:user@sip.domain.net.
CSeq: 22 REGISTER.
Call-ID: XmS5IC-57Z.
Max-Forwards: 70.
Supported: replaces, outbound, gruu.
Accept: application/sdp.
Accept: text/plain.
Accept: application/vnd.gsma.rcs-ft-http+xml.
Contact: <sip:user@192.0.2.1;transport=udp>;message-expires=300;+sip.instance="<urn:uuid:afdb1637-5e82-4375-a1ae-70dfdf9bbbf1>".
Expires: 0.
User-Agent: Linphone Desktop/4.2.4 (macOS 10.15, Qt 5.14.2) LinphoneCore/4.4.9-1-g18c59c7db.
Authorization:  Digest realm="sip.domain.net", nonce="5fec6f4100000000f9df525fee27a0b9aaa5e847c045b16a", username="user@sip.domain.net",  uri="sip:sip.domain.net", response="eb4c689165facc714912c98fb0b5cfcc".
</pre>

**Expected behavior**

`pv_www_authorize` should at least provide a way to check against userpart of username.

**Relevant System Logs**

**OS/environment information**
 - Operating System: `Debian 9`
 - OpenSIPS installation: `apt.opensips.org stretch/2.4-releases`
 - other relevant information:

**Additional context**

**Comment by bogdan-iancu** [comment]
@c-leroy , the `auth_get_ha1()` takes the username and password (for auth) from variables, not from the SIP messages. So, what do you populate into the `username_spec` at script level?

**Comment by c-leroy** [comment]
Here a snippet of configuration, for REGISTER messages :
```
modparam("auth_db", "calculate_ha1", no)
modparam("auth_db", "password_column", "ha1")
modparam("auth_db", "password_column_2", "ha1b")
modparam("auth_db", "load_credentials", "$avp(password)=ha1")
modparam("auth_db", "use_domain", 0)
modparam("auth", "username_spec", "$avp(username)")
modparam("auth", "password_spec", "$avp(password)")
modparam("auth", "calculate_ha1", no)

route{
    #...
    if (is_method("REGISTER")) {
      if ($au == null) {
        www_challenge("sip.domain.net", "0");
        exit;
      }
      cache_fetch("local","password_$au",$avp(password));
      if ($avp(password) == null || $avp(password) == "<null>") {
        www_authorize("sip.domain.net", "subscriber");
        $var(auth_rc)=$rc;
        cache_store("local","password_$au","$avp(password)",3600);
      } else {
        $avp(username)=$au;
        pv_www_authorize("sip.domain.net");
        $var(auth_rc)=$rc;
      }
      if ($var(auth_rc) < 0) {
        switch($var(auth_rc)) {
            case -1:
                xlog("L_NOTICE","C5::$avp(nat)::$proto:$si:$sp::$Au::$rm::403::Invalid User\n");
                sl_send_reply("403","Authentication failure");
                exit;
            break;
            case -2:
                xlog("L_NOTICE","C5::$avp(nat)::$proto:$si:$sp::$Au::$rm::403::Invalid Password\n");
                sl_send_reply("403","Authentication failure");
                exit;
            break;
            case -3:
                xlog("L_INFO","C5::$avp(nat)::$proto:$si:$sp::$Au::$rm::401::Stale Nonce\n");
            break;
            case -4:
                xlog("L_NOTICE","C5::$avp(nat)::$proto:$si:$sp::$Au::$rm::401::No Credentials\n");
            break;
            case -5:
                xlog("L_NOTICE","C5::$avp(nat)::$proto:$si:$sp::$Au::$rm::500::Generic Error\n");
                sl_send_reply("500","Internal server Error");
                exit;
            break;
            default:
               xlog("L_NOTICE","C5::$avp(nat)::$proto:$si:$sp::$Au::$rm::403::Return Code ($rc)\n");
                sl_send_reply("500","Internal server Error");
        }
        www_challenge("sip.domain.net", "0");
        exit;
      }
      #...
```

**Comment by bogdan-iancu** [comment]
I see where you are coming from with this :) 
Well, it may be considered an bug or not. The idea is that the auth layer needs the username only. In the auth_db, the module takes care to properly extract the username only. But for auth module, the script should take care of passing the username only to the auth layer. You may do:
```
 } else {
        if ( $(au{s.select,1,@})==NULL)
            $avp(username)=$au;
        else
             $avp(username)=$(au{s.select,0,@});
        pv_www_authorize("sip.domain.net");
        $var(auth_rc)=$rc;
      }
```

**Comment by c-leroy** [comment]
Thanks for your reply, I will try your conf.

According to https://www.opensips.org/Documentation/Script-CoreVar-2-4 :
```
3.4 Auth username user
$au - user part of username from Authorization or Proxy-Authorization header
```

**Comment by c-leroy** [comment]
I tried your conf mod, just to be sure : it does not work.

In fact, auth module checks against the whole field (`auth_get_ha1()`, line 382 of modules/auth/auth_mod.c) since commit 3b6685542b780ec47042a0e376de108e6d9c5fb7:
```C
    if(sval.rs.len!= _username->whole.len
        || strncasecmp(sval.rs.s, _username->whole.s, sval.rs.len))
```

**Comment by bogdan-iancu** [comment]
Thanks @c-leroy for the checking. Bottom line , the problem is with the check at line 382. If we expect all the time to receive a username-only in the `$avp(username)`, then we need to check it against the `_username->user` part only.

**Comment by bogdan-iancu** [comment]
ping @c-leroy

**Comment by c-leroy** [comment]
I don't know what to think about it.
I expected it would behave like auth_db, but it is not the case : auth_db can handle both cases (use_domain parameter), with optimizations (ha1 and ha1b pre-computation).
I have to settle and read the docs to understand how it would appear to a newcomer.

**Comment by sobomax** [comment]
@c-leroy @bogdan-iancu I can shed some light to this. I came to this old turd as well while working on RFC8760 auth patch. Basically db_auth has very old hack pre-dating RFC3261, which basically checks if username has @ and then checking auth with and without @host portion. This is the reason why we store two versions of md5 hash one called HA1 and the other HA1b.  As discussed with @liviuchircu this old hack is not something that RFC3261 requires and hardly useful now almost 20 years later. The plan was to get rid of it as part of the RFC8760 sweep which would hopefully make its way into 3.2. As such the best way to fix UAC configuration and don't rely on this being the case.

**Comment by bogdan-iancu** [comment]
@sobomax , the issue here is not about the ```auth_db```, but ```auth``` module (doing auth with credentials provided from script, not from DB). And the issue is a logical one, on how to properly compare the auth username you provide from script (which in 100% of cases is just a user part) with the auth username taken from the SIP auth hdrs (which may contain a domain part also).
My commit changes only the validation check, does not affect the actual auth part.

**Comment by bogdan-iancu** [comment]
@c-leroy , please check the fix I just pushed, hopefully it will address you issue

**Comment by c-leroy** [comment]
@bogdan-iancu I checked it : username verification succeeds now, but it fails to authenticate (invalid password).
I checked if the response from my softphone is correctly computed: it is.

**Comment by sobomax** [comment]
> @sobomax , the issue here is not about the `auth_db`, but `auth` module (doing auth with credentials provided from script, not from DB). And the issue is a logical one, on how to properly compare the auth username you provide from script (which in 100% of cases is just a user part) with the auth username taken from the SIP auth hdrs (which may contain a domain part also).
> My commit changes only the validation check, does not affect the actual auth part.

@bogdan-iancu well frankly this change is still of dubious utility. According to the RFC, there us no underlying structure in digest username, it's just one whole string. As such if someone wants to auth username of "foo@bar.com" both value in the username part and AVP should contain "foo@bar.com".  Period end of story. If I interpret your change correctly, it would allow AVP in this case to be just "foo" and totally ignore the @bar.com part. This is quite dangerous in my view since consider the situation when system has two valid accounts one with "foo" as username and the other is "foo@bar.com". With this new code we could either end up with false positive or false negative doing auth. As far as I can tell the issue with @c-leroy's scenario is that he has "user@sip.domain.net" as digest username, but wants to do auth as if it only had "user" part. The proper fix is to adjust configuration of the linphone to not append "@sip.domain.net" to username.

**Comment by sobomax** [comment]
P.S. I suspect this change would break legit users if the auth API who provide proper "user@sip.domain.net" in the AVP, since now you are going to run`strncasecmp("user@sip.domain.net", "user", strlen("user@sip.domain.net"))` and it's not going to work.

**Comment by c-leroy** [comment]
@sobomax We are a telco, and we have to cope with various UAC we do not manage.
We want to be able to authenticate users using cached credentials from the first call to auth_db, in both cases (username with or without domain).

**Comment by c-leroy** [comment]
As I expressed, there is no obvious response to this issue.
My main concern is "pv_www_authorize should at least provide a way to check against userpart of username" - like www_authorize does.
Other considerations are beyond my scope now... But I will understand these if they are well-documented.

**Comment by sobomax** [comment]
@c-leroy there is 100,500 ways to misconfigure SIP endpoint. Pushing a fix for what is essentially a misconfiguration at the edge to the SIP core is a good way to break things elsewhere. In your case I think you should be really be using "$aU", not "$au"!

```
3.4 Auth username user
$au - user part of username from Authorization or Proxy-Authorization header
...
3.13 Auth whole username
$aU - whole username from Authorization or Proxy-Authorization header
```

**Comment by c-leroy** [comment]
@sobomax What do we then set for load_credentials for auth_db ?
It is currently set to ha1.

**Comment by sobomax** [comment]
p.s. This macro should be deprecated. There is no "user" portion in the username.

```
3.4 Auth username user
$au - user part of username from Authorization or Proxy-Authorization header
```

**Comment by c-leroy** [comment]
I agree this "fix" will break a 7-year old behaviour. I'm not confident into putting it upstream...

**Comment by c-leroy** [comment]
There is a user part, real life tells us... or I am misleading

**Comment by sobomax** [comment]
@c-leroy No, there is not! You can configure your linphone just as well to put "user#sip.domain.net" into username or whatever.

**Comment by c-leroy** [comment]
As I told you, we are a telco and do not manage what are clients are doing.
BTW, I had to "hack' MY linphone to send such a username field...

**Comment by sobomax** [comment]
Just in case the very same username is also part of the hashing procedure according to RFC. As such if your UA uses "user" for that but "user@sip.domain.net" in SIP message the hash is not going to match anyways, unless the server applies the same trick with hash as well. Which is what auth_db does unfortunately.

**Comment by c-leroy** [comment]
Unfortunately, as you said. 👍

**Comment by sobomax** [comment]
@c-leroy try with '$aU'

**Comment by c-leroy** [comment]
And what about password stored in $avp(password) with auth_db? (It's ha1, staticly configured..)
I'll give it a try.

**Comment by sobomax** [comment]
auth_db should just work. I think you are using correct hash, it's just that you dropped @sip.domain.net along the way. Keep in mind you'd have to also undo @bogdan-iancus commit if my reading of the code is correct.

**Comment by c-leroy** [comment]
auth_db works, as it did before.
I tried with a previous build without this commit, and I end up with an incorrect password when checking with pv_www_authorize.
I still have :
modparam("auth_db", "calculate_ha1", no)
modparam("auth_db", "password_column", "ha1")
modparam("auth_db", "password_column_2", "ha1b")
modparam("auth_db", "load_credentials", "$avp(password)=ha1")
# use_domain désactivé pour compatibilité avec les PCSCF 2.0
modparam("auth_db", "use_domain", 0)
modparam("auth", "username_spec", "$avp(username)")
modparam("auth", "password_spec", "$avp(password)")
modparam("auth", "calculate_ha1", no)

**Comment by bogdan-iancu** [comment]
@sobomax thanks for your input here, you made a point - from RFC perspective, the auth username is opaque, it has no structure.

Throwing some light into `auth` versus `auth_db` - from checking the auth response, both do the same, they simply use the whole digest username, as received in SIP hdrs. Still, the `auth_db` is trying to see if the auth username is "user@domain", not for auth purposes, but in order to properly query the DB tables for fetching the user's passwd. The `auth` module does not have this problem as it gets the pwd from script and not via a DB query.
Bottom line, the change I did should be reverted and from script level, you should use the `$aU` and not `$au` . Do we agree on this ?

**Comment by sobomax** [comment]
> Bottom line, the change I did should be reverted and from script level, you should use the $aU and not $au . Do we agree on this ?

@bogdan-iancu we are on the same page, thanks! Moreover the $au variable should either be removed or amended with a massive warning saying that its use would result in non-RFC3261 behavior. I suspect lot of people out there are using it not realizing that once someone would try to use @ in username the auth would fail in weird ways....

**Comment by sobomax** [comment]
> @sobomax thanks for your input here, you made a point - from RFC perspective, the auth username is opaque, it has no structure.
> 
> Throwing some light into `auth` versus `auth_db` - from checking the auth response, both do the same, they simply use the whole digest username, as received in SIP hdrs. Still, the `auth_db` is trying to see if the auth username is "user@domain", not for auth purposes, but in order to properly query the DB tables for fetching the user's passwd. 

@bogdan-iancu I still think you are overlooking [HA1b code here](https://github.com/OpenSIPS/opensips/blob/master/modules/auth_db/authorize.c#L76):

```
	/* should we calculate the HA1, and is it calculated with domain? */
	if (_username->domain.len && !calc_ha1) {
		col[0] = &pass_column_2 ;
		CON_PS_REFERENCE(auth_db_handle) = &auth_ha1b_ps;
	} else {
		col[0] = &pass_column;
		CON_PS_REFERENCE(auth_db_handle) = &auth_ha1_ps;
	}
```

**Comment by sobomax** [comment]
P.S. We were hoping to remove it as part of the RFC8760 work, so that we won't have to support ha1b_sha256 & ha1b_sha512_256 https://github.com/sippy/opensips/commit/2ed9866479aa8f580e99b29abb9a39593bca7acc

**Comment by bogdan-iancu** [comment]
> > Bottom line, the change I did should be reverted and from script level, you should use the $aU and not $au . Do we agree on this ?
> 
> @bogdan-iancu we are on the same page, thanks! Moreover the $au variable should either be removed or amended with a massive warning saying that its use would result in non-RFC3261 behavior. I suspect lot of people out there are using it not realizing that once someone would try to use @ in username the auth would fail in weird ways....

You can count me in - I was also fooled by the $au (I thought it returns the exact "username" from the SIP hdr). Hoping that we are almost the only ones knowing the secret of $au versus $aU, I would suggest to have $au returning the whole username field and either drop $aU , either keep it as == to $au

**Comment by bogdan-iancu** [comment]
> > @sobomax thanks for your input here, you made a point - from RFC perspective, the auth username is opaque, it has no structure.
> > Throwing some light into `auth` versus `auth_db` - from checking the auth response, both do the same, they simply use the whole digest username, as received in SIP hdrs. Still, the `auth_db` is trying to see if the auth username is "user@domain", not for auth purposes, but in order to properly query the DB tables for fetching the user's passwd.
> 
> @bogdan-iancu I still think you are overlooking [HA1b code here](https://github.com/OpenSIPS/opensips/blob/master/modules/auth_db/authorize.c#L76):
> 
> ```
> 	/* should we calculate the HA1, and is it calculated with domain? */
> 	if (_username->domain.len && !calc_ha1) {
> 		col[0] = &pass_column_2 ;
> 		CON_PS_REFERENCE(auth_db_handle) = &auth_ha1b_ps;
> 	} else {
> 		col[0] = &pass_column;
> 		CON_PS_REFERENCE(auth_db_handle) = &auth_ha1_ps;
> 	}
> ```

@sobomax , IMO that is totally bogus. The auth computation should not care of any potential structuring of the username fields in auth response. The "b" version should be simply purged out.

**Comment by bogdan-iancu** [comment]
OK, I reverted the yesterday's commits, so @c-leroy you can try by using the `$aU` in cfg (for auth username) - it should work

@sobomax , if you agree, for 3.2, I will remove all the user@domain madness for auth username, keeping only `$au` for the (whole) username and strip out the HA1b stuff

**Comment by c-leroy** [comment]
I tried in comment https://github.com/OpenSIPS/opensips/issues/2356#issuecomment-760528225 : no success.
If I manually set `$avp(password)` with ha1b (hard-coded in configuration script), it succeeds.

**Comment by sobomax** [comment]
> I tried in comment [#2356 (comment)](https://github.com/OpenSIPS/opensips/issues/2356#issuecomment-760528225) : no success.
> If I manually set `$avp(password)` to ha1b value before calling `pv_www_authorize()` (hard-coded in configuration script for this user), it succeeds.

@c-leroy which means your endpoint is not computing MD5 hash correctly (i.e. using full username value). The correct workaround in this case would be to store some extra boolean flag as part of the auth structure in the DB and use either ha1 or ha1b depending on its value. Hope it helps!

**Comment by c-leroy** [comment]
I'm quite sure it does. I can send you the challenge, which I checked with a script, if you give me a secure way to do this.
I know username is "opaque", so does my response check script does...

**Comment by sobomax** [comment]
@c-leroy but what you are saying makes no sense, I am sorry. ha1b is exactly that - HA1 value but computed with only "user" portion of the username. As such, if your auth succeds with ha1b but not with ha1, this effectively means your endpoint is violating RFC3261. There is no other way around it. The verification should be trivial run the following two commands:

```
$ username="user@sip.domain.net";realm="sip.domain.net";password="abcdefg67"; echo -n "${username}:${realm}:${password}" | md5sum
$ username="user";realm="sip.domain.net";password="abcdefg67"; echo -n "${username}:${realm}:${password}" | md5sum
```
(replace abcdefg67 with actual password) and then compare which HA1 works. If it's the former your endpoing is RFC3261-compliant, if the second - it's not.

**Comment by sobomax** [comment]
P.S. Your $avp(password) is populated here `modparam("auth_db", "load_credentials", "$avp(password)=ha1")`, so it's not very "arbitrary". Nobody prevents you from doind something like modparam("auth_db", "load_credentials", "$avp(password)=ha1;$avp(passwordB)=ha1b";$avp(use_planB)=broken_auth) and then use either password or passwordB sparingly based on use_planB value.

**Comment by c-leroy** [comment]
> @c-leroy but what you are saying makes no sense, I am sorry. ha1b is exactly that - HA1 value but computed with only "user" portion of the username. As such, if your auth succeds with ha1b but not with ha1, this effectively means your endpoint is violating RFC3261. There is no other way around it. The verification should be trivial run the following two commands:
> 
> ```
> $ username="user@sip.domain.net";realm="sip.domain.net";password="abcdefg67"; echo -n "${username}:${realm}:${password}" | md5sum
> $ username="user";realm="sip.domain.net";password="abcdefg67"; echo -n "${username}:${realm}:${password}" | md5sum
> ```
> 
> (replace `abcdefg67 ` with actual password) and then compare which HA1 works. If it's the former your endpoint is RFC3261-compliant (which I seriously doubt), if the second - it's not.

ha1b is computed for whole username field.



> P.S. Your $avp(password) is populated here `modparam("auth_db", "load_credentials", "$avp(password)=ha1")`, so it's not very "arbitrary". Nobody prevents you from doind something like `modparam("auth_db", "load_credentials", "$avp(password)=ha1;$avp(passwordB)=ha1b;$avp(use_planB)=broken_auth")` instead and then use either `password` or `passwordB` sparingly by assigning password=passwordB conditionally before you call pv_www_authorize() based on `use_planB` value, which corresponds to the field `broken_auth` in the auth DB. You can even do without broken_auth flag (similarly to what auth_db does), by first trying to auth with `password` and if that fails with `passwordB`. The reason we've got this hack in auth_db module is partially because there were no such possibility in 2003, scripting was much more primitive back then.

Thanks! I missed the ability to load both columns.

```
modparam("auth_db", "calculate_ha1", no)
modparam("auth_db", "password_column", "ha1")
modparam("auth_db", "password_column_2", "ha1b")
modparam("auth_db", "load_credentials", "$avp(ha1)=ha1;$avp(ha1b)=ha1b")
modparam("auth_db", "use_domain", 0)
modparam("auth", "username_spec", "$avp(username)")
modparam("auth", "password_spec", "$avp(password)")
modparam("auth", "calculate_ha1", no)
#...
      if ($au == null) {
        www_challenge("sip.domain.net", "0");
        exit;
      }
      cache_fetch("local","password_$au",$avp(password));
      if ($avp(password) == null || $avp(password) == "<null>") {
        www_authorize("sip.domain.net", "subscriber");
        $var(auth_rc)=$rc;
        if ($(aU{s.select,1,@})==NULL) {
          $avp(password) = $avp(ha1);
        } else {
          $avp(password) = $avp(ha1b);
        }
        cache_store("local","password_$au","$avp(password)",3600);
      } else {
        $avp(username)=$aU;
        pv_www_authorize("sip.domain.net");
        $var(auth_rc)=$rc;
      }
      if ($var(auth_rc) < 0) {
```

**Comment by bogdan-iancu** [comment]
@c-leroy , are we good here, ready for closing it?

**Comment by c-leroy** [comment]
OK for me!

**Comment by c-leroy** [comment]
I updated https://github.com/OpenSIPS/opensips/issues/2356#issuecomment-762972874 to only use $aU (it still relied on current $au value). If you implement your idea from https://github.com/OpenSIPS/opensips/issues/2356#issuecomment-760986024, we will have to simply replace $aU with $au.
Note: about "we are almost the only ones knowing the secret of $au versus $aU", it is documented in https://www.opensips.org/Documentation/Script-CoreVar-2-4#toc7 and https://www.opensips.org/Documentation/Script-CoreVar-2-4#toc16
Do not forget to warn people if you intend to change these variables.
